### PR TITLE
Context-ified file IO

### PR DESCRIPTION
**Pull request by QuLogic**
This is mostly independent, but there are a couple of small conflicts, so I based it on the previous PR, which makes it look a little big...

This ensures files are correctly closed by opening them in a `with` context. In Python 3, it will cause a `ResourceWarning` if you don't close files explicitly or with a context manager.

### CODE REVIEW

**Comment by QuLogic** [comment]
So, tests will fail because of the `MagicMock`, but I'm not entirely sure where the expected result originates, or I would fix it myself.

**Comment by rhattersley** [comment]
> In Python 3, it will cause a ResourceWarning if you don't close files explicitly or with a context manager.

NB. Only if you fiddle with the warning settings (which nose does):

```python
$ python
>>> def test():
...  open('foo', 'w')
... 
>>> test()
>>> 
```

```python
$ python -Wdefault
>>> def test():
...  open('foo', 'w')
... 
>>> test()
__main__:2: ResourceWarning: unclosed file <_io.TextIOWrapper name='bar' mode='w' encoding='UTF-8'>
>>> 
```

**Comment by rhattersley** [comment]
The file-specific changes look good - thanks. Looks like this just needs other PRs merging and/or a rebase and then I'm :+1:

**Comment by QuLogic** [comment]
Rebased this one.

File: docs/iris/src/developers_guide/gitwash_dumper.py [hunk 1]
```diff
@@ -57,16 +57,19 @@ def filename_search_replace(sr_pairs, filename, backup=False):
     ''' Search and replace for expressions in files
 
     '''
-    in_txt = open(filename, 'rt').read(-1)
+    with open(filename, 'rt') as in_fh:
```
Comment (pelson) [inline docs/iris/src/developers_guide/gitwash_dumper.py]: Incidentally, this isn't our code. See https://github.com/matthew-brett/gitwash/blob/master/gitwash_dumper.py.

**Comment by pelson** [comment]
I also had a go at fixing this once upon a time, but my approach was to take file handles as arguments, rather than filenames. It was a challenge in some situations, and ended up in me writing a library to resuscitate closed file handles (where possible), but I did get 80% of the way there.
Did you consider this approach?

**Comment by QuLogic** [comment]
Not sure what you mean there.

File: docs/iris/example_code/General/custom_file_loading.py [hunk 2]
```diff
@@ -63,74 +70,84 @@ def load_NAME_III(filename):
...
+        headers = {}
+
+        # Skip the NAME header of the file which looks something like
+        # 'NAME III (version X.X.X)'.
+        next(file_handle)
+
+        # Read the next 16 lines of header information, putting the form
+        # "header name:    header value" into a dictionary.
+        for _ in range(16):
+            header_name, header_value = file_handle.next().split(':')
```
Comment (rhattersley) [inline docs/iris/example_code/General/custom_file_loading.py]: This needs to be `... = next(file_handle).split(':')` to work on Python 3.

File: docs/iris/example_code/General/custom_file_loading.py [hunk 3]
```diff
@@ -63,74 +70,84 @@ def load_NAME_III(filename):
...
-
-    # convert the time to python datetimes
-    new_time_column_header = []
-    for i, t in enumerate(column_headings['time']):
-        # the first 4 columns aren't time at all, so don't convert them to datetimes
-        if i >= 4:
-            new_time_column_header.append(datetime.datetime.strptime(t, UTC_format))
-        else:
+    # Loading a file gives a generator of lines which can be progressed using
+    # the next() method. This will come in handy as we wish to progress through
```
Comment (rhattersley) [inline docs/iris/example_code/General/custom_file_loading.py]: "the next() method" ... should probably now read "the next() function"

File: docs/iris/example_code/General/custom_file_loading.py [hunk 4]
```diff
@@ -63,74 +70,84 @@ def load_NAME_III(filename):
...
+            headers[header_name] = header_value
+
+        # Skip the next blank line in the file.
+        next(file_handle)
+
+        # Read the next 7 lines of column definitions.
+        column_headings = {}
+        for column_header_name in COLUMN_NAMES:
+            column_headings[column_header_name] = [
+                col.strip() for col in file_handle.next().split(',')
```
Comment (rhattersley) [inline docs/iris/example_code/General/custom_file_loading.py]: As above, this needs to be `next(file_handle)`

File: lib/iris/fileformats/ff.py [hunk 5]
```diff
@@ -650,107 +650,108 @@ def _extract_field(self):
...
-                field.bplat = grid.pole_lat
-                field.bplon = grid.pole_lon
-            elif no_x or no_y:
-                warnings.warn('Partially missing X or Y coordinate values.')
-
-            # Check for LBC fields.
+        with open(self._ff_header.ff_filename, 'rb') as ff_file:
+            ff_file_seek = ff_file.seek
+
             is_boundary_packed = self._ff_header.dataset_type == 5
```
Comment (rhattersley) [inline lib/iris/fileformats/ff.py]: (Note to self: this line appears twice in the original code (i.e. it's a "consequence-free" bug) which is why the code diff gets split up here)

File: docs/iris/src/developers_guide/gitwash_dumper.py (hunk 1 above)
Comment (rhattersley) [inline docs/iris/src/developers_guide/gitwash_dumper.py]: I don't mind what happens here - we can keep the changes or not. As an aside, it'd be nice to push them upstream either way.

File: docs/iris/example_code/General/custom_file_loading.py (hunk 2 above)
Comment (rhattersley) [inline docs/iris/example_code/General/custom_file_loading.py]: I see you've already [sorted this out](https://github.com/QuLogic/iris/tree/py3k-builtin-changes) in a [subsequent commit](https://github.com/QuLogic/iris/commit/067fd2e037bad20729f7e4d64f910add0b41ccf0). I'm perfectly happy to wait for that. :smile:

**Comment by rhattersley** [comment]
@QuLogic & @pelson - please feel free to back out the gitwash changes, but I'd rather not hold up all the other good work waiting in the wings while we decide what to do about it.

**Comment by pelson** [comment]
:+1: - thanks @QuLogic and @rhattersley

File: docs/iris/src/developers_guide/gitwash_dumper.py (hunk 1 above)
Comment (QuLogic) [inline docs/iris/src/developers_guide/gitwash_dumper.py]: I've opened a [PR](https://github.com/matthew-brett/gitwash/pull/7) upstream.

**Pull request merged**